### PR TITLE
12/14 article: test2doc

### PR DESCRIPTION
**Pull request by adams-sarah**
This is a little early, but here's my article on test2doc scheduled for 12/14.

Cheers!
Sarah

### CODE REVIEW

**Comment by davecheney** [comment]
You never need to apologise for getting your article ready _early_, Sarah :)

**Comment by adams-sarah** [comment]
Haha, right. Thanks Dave.

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+    - URI
+    - query params
+    - request headers
+    - request body
+- response format:
+    - status code
+    - response headers
+    - response body
+
+<br>
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: Just curious what these <br> tags are in here for?

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+```
+
+And a test for this handler func, eg.:
+
+```go
+func TestGetWidget(t *testing.T) {
+    urlPath := fmt.Sprintf("/widgets/%d", 2)
+
+    resp, err := http.Get(server.URL + urlPath)
+    // assert all the things...
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: x-all-the-things-template

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+}
+```
+
+Test2doc will automatically generate markdown documentation for this 
+endpoint in the [API 
+Blueprint](https://github.com/apiaryio/api-blueprint/blob/master/API%20B
+lueprint%20Specification.md) format as your tests run, like so:
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: broken link: extra + sign in B+lueprint

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+## Getting started with test2doc
+A big goal of mine as I was writing this project was to limit the 
+amount of additional code needed to get test2doc up and running.
+
+Requirements:
+
+1. You must have a `TestMain` for each of the packages you wish to 
+document.
+2. All of the tests in your package must share a single `test.Server` 
+instance (from test2doc/test package)
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: I'd like to see a link to https://godoc.org/github.com/adams-sarah/test2doc/test#Server or https://godoc.org/github.com/adams-sarah/test2doc/test

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// .. then run your tests as usual
+	exitCode := m.Run()
+
+
+	// 3. Finally, you must tell the wrapped server when you are done testing
+	//    so that the buffer can be flushed to an apib doc file
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: apib?

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+	server.Finish()
+
+	// note that os.Exit does not respect defers.
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: yikes. I never knew that.

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+```go
+type ResponseWriter struct {
+	HandlerInfo HandlerInfo
+	URLVars     map[string]string
+	W           *httptest.ResponseRecorder
+}
+```
+
+The test2doc `ResponseWriter`'s implementations of the `Header` and
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: note: this sentence took me a few reads, but I got it.

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+package) until it finds a caller whose declaration is inside the 
+package we are testing (in all likelihood, the handler).
+
+Once we have the handler's function name, we can get the Go doc string 
+using the `go/doc` package.
+
+
+## Thoughts for the future
+I'd like to convert the main types to interfaces, allowing support for 
+formats other than the API Blueprint format (eg, Google's discovery js).
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: is Google's discovery js this? https://developers.google.com/discovery/libraries?hl=en

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+I'd also like to improve upon the "handler-finding" algorithm (above), 
+to make it more reliably accurate.
+
+## Contributions welcome!
+Drop me a line at sadams.codes@gmail.com, or adams-sarah on github.
```
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: a link to https://github.com/adams-sarah would be nice :-)

**Comment by nathany** [comment]
Thanks Sarah. It's great to see your project taking shape.

File: content/advent-2015/test2doc-generate-your-api-docs.md (hunk 1 above)
Comment (adams-sarah) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: I put them in for formatting/readability. There isn't enough spacing between sections by default IMO.
I can take them out if you prefer, just let me know.

Comment (adams-sarah) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: The `<br>` tag you're referring to in this particular case evens out the spacing before and after the list:
screen shot 2015-12-10 at 10 14 24 am

File: content/advent-2015/test2doc-generate-your-api-docs.md (hunk 5 above)
Comment (adams-sarah) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: hm, yea. "apib" is the extension for an API Blueprint file.
How about "...flushed to an API Blueprint doc file"?

File: content/advent-2015/test2doc-generate-your-api-docs.md (hunk 8 above)
Comment (adams-sarah) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: To be honest, I have never actually played with Google discovery.js. I recently got a question on twitter asking if I'll support discovery.js in the future, which is why "discover js" is in here. 
I'm going to take out this reference, and leave it at "formats other than the API Blueprint format"

File: content/advent-2015/test2doc-generate-your-api-docs.md (hunk 7 above)
Comment (adams-sarah) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: Updated to: 
"Test2doc's `ResponseWriter` implements the `Header` and 
`WriteHeader` methods by just falling back those of its 
embedded `httptest.ResponseRecorder`"

File: content/advent-2015/test2doc-generate-your-api-docs.md (hunk 1 above)
Comment (nathany) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: Okay. I was just curious.

File: content/advent-2015/test2doc-generate-your-api-docs.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+```go
+type ResponseWriter struct {
+	HandlerInfo HandlerInfo
+	URLVars     map[string]string
+	W           *httptest.ResponseRecorder
+}
+```
+
+Test2doc's `ResponseWriter` implements the `Header` and 
+`WriteHeader` methods by just falling back those of its
```
Comment (sbinet) [inline content/advent-2015/test2doc-generate-your-api-docs.md]: `s/back those/back to those` ?

**Comment by adams-sarah** [comment]
Hey, @nathany, just curious about the status of this article. Is it going to go out?

**Pull request merged**